Detectors/ITSMFT/ITS/tracking/src/Tracker.cxx

```diff
@@ -273,151 +266,94 @@ void Tracker::findTracks(const ROframe& event)
-#ifdef CA_DEBUG
-    mDebugger->dumpTrackToBranchWithInfo("testBranch", temporaryTrack, event, mPrimaryVertexContext, true);
-#endif
+    fitSuccess = fitTrack(temporaryTrack, mTrkParams[0].NLayers - 1, -1, -1, mTrkParams[0].FitIterationMaxChi2[1]);
     if (!fitSuccess) {
       continue;
     }
-    CA_DEBUGGER(refitCounters[nClusters - 4]++);
+    // temporaryTrack.setROFrame(rof);
     tracks.emplace_back(temporaryTrack);
-    CA_DEBUGGER(assert(nClusters == temporaryTrack.getNumberOfClusters()));
   }
   //mTraits->refitTracks(event.getTrackingFrameInfo(), tracks);
 
   std::sort(tracks.begin(), tracks.end(),
             [](TrackITSExt& track1, TrackITSExt& track2) { return track1.isBetter(track2, 1.e6f); });
 
-#ifdef CA_DEBUG
-  // std::array<int, 26> sharingMatrix{0};
-  // int prevNclusters = 7;
-  // auto cumulativeIndex = [](int ncl) -> int {
-  //   constexpr int idx[5] = {0, 5, 11, 18, 26};
-  //   return idx[ncl - 4];
-  // };
-  // std::array<int, 4> xcheckCounters{0};
-#endif
-
   for (auto& track : tracks) {
-    CA_DEBUGGER(int nClusters = 0);
     int nShared = 0;
     for (int iLayer{0}; iLayer < mTrkParams[0].NLayers; ++iLayer) {
       if (track.getClusterIndex(iLayer) == constants::its::UnusedIndex) {
         continue;
       }
-      nShared += int(mPrimaryVertexContext->isClusterUsed(iLayer, track.getClusterIndex(iLayer)));
-      CA_DEBUGGER(nClusters++);
+      nShared += int(mTimeFrame->isClusterUsed(iLayer, track.getClusterIndex(iLayer)));
     }
 
-    // #ifdef CA_DEBUG
-    //     assert(nClusters == track.getNumberOfClusters());
-    //     xcheckCounters[nClusters - 4]++;
-    //     assert(nShared <= nClusters);
-    //     sharingMatrix[cumulativeIndex(nClusters) + nShared]++;
-    // #endif
-
     if (nShared > mTrkParams[0].ClusterSharing) {
       continue;
     }
 
-    // #ifdef CA_DEBUG
-    //     nonsharingCounters[nClusters - 4]++;
-    //     assert(nClusters <= prevNclusters);
-    //     prevNclusters = nClusters;
-    // #endif
-
+    std::pair<int, int> rofs[3]{{-1, 0}, {-1, 0}, {-1, 0}};
     for (int iLayer{0}; iLayer < mTrkParams[0].NLayers; ++iLayer) {
       if (track.getClusterIndex(iLayer) == constants::its::UnusedIndex) {
         continue;
       }
-      mPrimaryVertexContext->markUsedCluster(iLayer, track.getClusterIndex(iLayer));
+      mTimeFrame->markUsedCluster(iLayer, track.getClusterIndex(iLayer));
+      int currentROF = mTimeFrame->getClusterROF(iLayer, track.getClusterIndex(iLayer));
+      for (int iR{0}; iR < 3; ++iR) {
+        if (rofs[iR].first == -1) {
+          rofs[iR].first = currentROF;
+        }
+        if (rofs[iR].first == currentROF) {
+          rofs[iR].second++;
+          break;
+        }
+      }
     }
-    mTracks.emplace_back(track);
+    if (rofs[2].first != -1) {
+      continue;
+    }
+    int trackROF{rofs[0].second > rofs[1].second ? rofs[0].first : rofs[1].first};
+    mTimeFrame->getTracks(trackROF).emplace_back(track);
```

Comment:
@mpuccio  For me, it breaks at this line 344.  Because, trackROF=-1.  I am trying to reconstruct 1000 pythia6 events.

Reply:
Fixed with the last commit! Thanks a lot for sharing the data